File: CPP/No121.cc
/**
 * Created by Xiaozhong on 2020/8/7.
 * Copyright (c) 2020/8/7 Xiaozhong. All rights reserved.
 */
#include "iostream"
#include "vector"

using namespace std;

class Solution {
public:
    int maxProfit(vector<int> &prices) {
        int n = prices.size();
        if (!n) return 0;
        int dp = 0;
        int min_price = prices[0];
        for (int i = 1; i < n; i++) {
            min_price = min(min_price, prices[i]);
            dp = max(dp, prices[i] - min_price);
        }
        return dp;
    }
};

int main(int argc, char **args) {
    Solution s;
    vector<int> prices = {7, 1, 5, 3, 6, 4};
    cout << s.maxProfit(prices);
}

File: Java/src/No121.java
/*
 * Copyright (c) 8/30/2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

public class No121 {
    public int maxProfit(int[] prices) {
        int n = prices.length, min = Integer.MAX_VALUE;
        int[] dp = new int[n + 1];
        for (int i = 0; i < n; ++i) {
            min = Math.min(min, prices[i]);
            dp[i + 1] = Math.max(dp[i], prices[i] - min);
        }
        return dp[n];
    }

    public static void main(String[] args) {
        int[] prices = {7, 6, 4, 3, 1};
        No121 no121 = new No121();
        System.out.println(no121.maxProfit(prices));
    }
}
